### ISSUE TITLE
Creating a class from scratch with diamond operator 

### ISSUE THREAD

**Issue by ufarooq** [Sable/soot#1225]
Hi, 

I am trying create a class from scratch using jimple APIs, the class should extend a super class by specifying a type parameter using diamond(``<T>``) operator. The class should look like this if we were writing in java. 

``` java
public class MyClassType extends TypeToken<MyClass> {

}
```

I can see ``` setSuperclass``` API for ```SootClass``, but I could not find any option to specify type parameter. 

Any idea on this?

Bests,
Umar

**Comment by mbenz89** [comment]
Hi Umar.

There are no type parameters in bytecode and, thus, not in Jimple. Java will keep objects specified with the generic type as if they were instances of its specified type (In your case this is Object since `MyClass` has no more specific type).

I suggest to create a class with generics and have a look at the bytecode (javap) or the jimple code to get a glimpse of that.

Best,
Manuel

**Comment by ufarooq** [comment]
Thanks @mbenz89 for the reply! I agree when example code is compiled and disassembled in jimple type ``MyClass`` is lost. 

So just to test type parameters, I am creating a simple class. 

```java
public class Test {
    private List<String> list;
    void test() {
        list = new ArrayList<String>();
        list.add("test");
    }
}
``` 
Now, when I use `` javap -p Test.class``, I see following output which provides type parameters as well. (``private java.util.List<java.lang.String> list;``)

```
Compiled from "Test.java"
public class Test {
  private java.util.List<java.lang.String> list;
  public Test();
  void test();
}
```
Next, Jimple does not inlcude type information and output looks like this: 

```
public class Test extends java.lang.Object
{
    private java.util.List list;

    public void <init>()
    {
        Test r0;

        r0 := @this: Test;

        specialinvoke r0.<java.lang.Object: void <init>()>();

        return;
    }

    void test()
    {
        Test r0;
        java.util.ArrayList $r1;
        java.util.List $r2, $r3;

        r0 := @this: Test;

        $r1 = new java.util.ArrayList;

        specialinvoke $r1.<java.util.ArrayList: void <init>()>();

        r0.<Test: java.util.List list> = $r1;

        $r2 = r0.<Test: java.util.List list>;

        interfaceinvoke $r2.<java.util.List: boolean add(java.lang.Object)>("test");

        return;
    }
}
```

Clearly, Either Soot drops such information during disassembly or APIs to access type parameters are missing. 

Bests,
Umar

**Comment by mbenz89** [comment]
You are right. In the early days of Java, generic type information was not stored in the bytecode. At some point, this information was annotated to the bytecode. Soot does not extract it, though. It shouldn't be too hard to also extract this information if you need it.

**Comment by ufarooq** [comment]
Manuel, Yes I need this information since it is an important part of my analysis. It would be great if you point out some most relevant classes to extract this information.

Best,
Umar

**Comment by ufarooq** [comment]
The signature tag for SootField contains such information. Which is sufficient to parse and extract the information I need for now. 

Still, someone who wants to create a class with the Type parameter will be unable to do so.

**Comment by ericbodden** [comment]
Thanks for the clarification @ufarooq. I doubt we will change this for Soot because it would likely not be backward compatible but we could keep this on the agenda for the next major release (FutureSoot) maybe. /cc @linghuiluo

**Comment by StevenArzt** [comment]
It would be less invasive to add a few convenience functions to handle the tags. Analyses that are aware of this infrastructure could then simply ask for the generic types. For example, we could have some ``RefType.getGenericType()`` method that checks for a tag and parses it. Analyses can then opt to query that or not.

**Comment by linghuiluo** [comment]
Thanks for the information. We will consider this for the next major release.
